feat(AIP-193): specify how to "change" messages.

shwoodard opened the pull request:
Introduces the concept that `LocalizedMessage` can be used to include a second error message aside from the debug message on `Status`. This message can be updated ongoingly. Further, explains how to use the `metadata` `map<string, string>` on `ErrorInfo` for dynamic variables found in error messages and otherwise.

Fixes: #1096

shwoodard committed:
feat(AIP-193): specify how to "change" messages.

Introduces the concept that `LocalizedMessage` can be used to include a
second error message aside from the debug message on `Status`. This
message can be updated ongoingly. Further, explains how to use the
`metadata` `map<string, string>` on `ErrorInfo` for dynamic variables
found in error messages and otherwise.

Fixes: #1096

shwoodard committed:
Edits

shwoodard committed:
Fix typo

shwoodard committed:
the section is immediately after now

shwoodard committed:
considerable editing

shwoodard committed:
Minor edits

shwoodard committed:
minor edits to rationale

shwoodard commented on aip/general/0193.md:
```diff
@@ -62,11 +62,90 @@ a [`PreconditionFailure`][PreconditionFailure].
+  ([citation](https://cloud.google.com/apis/design/errors#error_model))
+
+Once launched the value of `message` at the root-level of `Status`
+**must not** change for the error payload to be backwards compatible.
```
@jskeet points out that this statement may need further clarification, e.g. definition around what "backward compatible" means in this context.

noahdietz replied:
I would also say we need clarification around what the "error payload" constitutes e.g. the `message` is paired with the `code` OR the `message` is paired with the specific error mode (e.g. invalid location errors), so that services can reasonably keep track of what `message` an "error payload" should use.

Not a blocking comment for this PR.

shwoodard replied:
Good point. However, this statement was removed, which is not to say that we do not need to address this concern at a later date.

noahdietz commented on aip/general/0193.md:
```diff
@@ -62,11 +62,90 @@ a [`PreconditionFailure`][PreconditionFailure].
+1. The payload **must** contain an error message in the `message` field
+   found at the root of [`Status`][Status].
+1. The payload **should** contain
+   [`google.rpc.LocalizedMessage`][LocalizedMessage] in `details` on
```
```suggestion
   [`google.rpc.LocalizedMessage`][LocalizedMessage] in the `details` field of
```
IDK I think it parallels with the language in the previous item for `message`.

shwoodard replied:
Structure changed after major edits.

noahdietz commented on aip/general/0193.md:
```diff
@@ -62,11 +62,90 @@ a [`PreconditionFailure`][PreconditionFailure].
+   [`google.rpc.LocalizedMessage`][LocalizedMessage] in `details` on
+   `Status`. This is the location of the second error message.
+
+The error message on `Status` should,
```
Is this a requirement word "should" or best practices "should"? 

Perhaps this:
> The error message `Status` is most effective when it,
>
> - is developer-facing, ...
> - explains the error...

OR make it a `**should**`

noahdietz commented on aip/general/0193.md:
```diff
@@ -62,11 +62,90 @@ a [`PreconditionFailure`][PreconditionFailure].
+#### Changing error messages
+
+A new, additional error message **may** be added because the error
+message at the root of the payload **must not** change once launched.
```
```suggestion
message at the root of the payload cannot be changed once launched.
```
I think this guidance is already provided and here it seems to be the motivating reason, not guidance.

I don't feel strongly about this comment though.

noahdietz commented on aip/general/0193.md:
```diff
@@ -62,11 +62,90 @@ a [`PreconditionFailure`][PreconditionFailure].
+improvement. It is potentially displayed directly to end users by
+client application.
+
+When adding a new error messsage with `LocalizedMessage`, `ErrorInfo`
```
```suggestion
When adding a new error messsage via `LocalizedMessage`, `ErrorInfo`
```
Nnit: Maybe more clear?

shwoodard replied:
Fixed.

noahdietz commented on aip/general/0193.md:
```diff
@@ -62,11 +62,90 @@ a [`PreconditionFailure`][PreconditionFailure].
+improvement. It is potentially displayed directly to end users by
+client application.
+
+When adding a new error messsage with `LocalizedMessage`, `ErrorInfo`
```
```suggestion
When adding a new error message with `LocalizedMessage`, `ErrorInfo`
```
Typo

shwoodard replied:
Fixed.

noahdietz commented on aip/general/0193.md:
```diff
@@ -62,11 +62,90 @@ a [`PreconditionFailure`][PreconditionFailure].
+
+When adding a new error messsage with `LocalizedMessage`, `ErrorInfo`
+**must** be introduced before or concurrently. If there are dyanmic
+variables found in the message, these variables must be included in the
```
In which message? The LocalizedMessage.message i.e. the "new" message? or the "old" one?

shwoodard replied:
I clarified this by saying that `ErrorInfo` is required for new errors and that variables **must** be in `metadata`. Feel free to reopen.

noahdietz commented on aip/general/0193.md:
```diff
@@ -62,11 +62,90 @@ a [`PreconditionFailure`][PreconditionFailure].
+
+When adding a new error messsage with `LocalizedMessage`, `ErrorInfo`
+**must** be introduced before or concurrently. If there are dyanmic
+variables found in the message, these variables must be included in the
```
```suggestion
variables found in the message, these variables **must be** included in the
```

shwoodard replied:
Fixed.

noahdietz commented on aip/general/0193.md:
```diff
@@ -62,11 +62,90 @@ a [`PreconditionFailure`][PreconditionFailure].
+
+**Warning:** If `LocalizedMessage` is released without `ErrorInfo` even
+the new error message **must not** change. To do so, would be backwards
+incompatible, effectively locking all error messages in the payload.
```
I'm not really sure what this is saying. Is this trying to say that "if `LocalizedMessage` is released, but `ErrorInfo` isn't at all, then the `LocalizedMessage.message` must match the original `Status.message`"?

shwoodard replied:
Rephrased.

noahdietz commented on aip/general/0193.md:
```diff
@@ -62,11 +62,90 @@ a [`PreconditionFailure`][PreconditionFailure].
+If the service is location aware, the value of `locale` **should**
+change.  See, "[Localization](#localization)". Otherwise, the field
+**may** contain the service's default locale, e.g.  "en-US", as its
+value, always, allowing the opportunity to add localization later.
```
Are the `**may**` and `always` conflicting here? Should it be `**must** contain the service's default locale...as its value`?

shwoodard replied:
Fixed.

noahdietz commented on aip/general/0193.md:
```diff
@@ -62,11 +62,90 @@ a [`PreconditionFailure`][PreconditionFailure].
+> The Book, "The Great Gatsby", is unvailable at the Library, "Garfield
+> East". It is expected to be available again on 2199-05-13.
+
+**Note:** The preceeding error message is made actionable by the
```
Is this really a "Note"? Kind of seems like a direct explanation of how the example fulfills the opening statement.

shwoodard replied:
Fixed.

noahdietz commented on aip/general/0193.md:
```diff
@@ -62,11 +62,90 @@ a [`PreconditionFailure`][PreconditionFailure].
+All dynamic variables found in error messages **must** also appear in
+the `map<string, string>` field, `metadata` (found on the *required*
+`ErrorInfo`).  For example, the `metadata` map for the sample error
+message above **must** include *at least* the following key/value pairs:
```
I'm not sure that an example should contain a requirement word. Perhaps replace it with `will`

shwoodard replied:
Fixed.

noahdietz commented on aip/general/0193.md:
````diff
@@ -62,11 +62,90 @@ a [`PreconditionFailure`][PreconditionFailure].
+message above **must** include *at least* the following key/value pairs:
+
+```yaml
+bookTitle: "The Great Gatsby"
````
So the metadata keys must be camelCase (JSON style)? Or is it based on the transport? e.g. ESF will coerce them to camelCase when sending back an HTTP response from a gRPC-HTTP transcoded Status, while pure gRPC would use snake_case?

idc either way but let's be specific about that.

shwoodard replied:
It must be camelCase. This is specified in the comment about `ErrorInfo`. Should it be specified here as well?

shwoodard replied:
Fixed.

noahdietz replied:
For sure thank you, this seems like a good spot to include it given the broad audience it will have.

noahdietz commented on aip/general/0193.md:
```diff
@@ -62,11 +62,90 @@ a [`PreconditionFailure`][PreconditionFailure].
+
+The error message, found in the payload on `Status`, **must** be in
+English. If a localized error message is also required, the service
+**should** use [`google.rpc.LocalizedMessage`][LocalizedMessage] in the
```
Not `**must**`?

shwoodard replied:
That was as it was before. I just moved the section so it looks new. Should I change it?

noahdietz replied:
Perhaps not if it was there before...kind of surprised it wasn't a must before.

noahdietz commented on aip/general/0193.md:
```diff
@@ -107,6 +186,16 @@ enables clients to pull such information programatically--without
 parsing the string error message. For these reasons, `ErrorInfo` is
 required for *every* error response.
 
+The field, `message`, on `Status` is immutable because clients **may**
```
Again, I'm not sure the requirement word style for `**may**` is right here since this is a reason for doing something.

shwoodard replied:
Removed additions to rationale section altogether since we are reversing the decision that `message` should be immutable.

sir-buckyball commented on aip/general/0193.md:
```diff
@@ -62,11 +62,90 @@ a [`PreconditionFailure`][PreconditionFailure].
+The error message on `Status` should,
+
+- Be developer-facing, human-readable, and in English
+- Both expalin the error and offer an actionable resolution to it
```
s/expalin/explain/

sir-buckyball commented on aip/general/0193.md:
```diff
@@ -62,11 +62,90 @@ a [`PreconditionFailure`][PreconditionFailure].
+#### Changing error messages
+
+A new, additional error message **may** be added because the error
+message at the root of the payload **must not** change once launched.
```
IMO changing the error message is not a breaking change and should not be treated as such. A well behaved client should operate based on error codes and not the contents of the messages themselves. If a client need the specificity of the error message itself for processing purposes, that's an indication that an error code or sub-code is needed.

shwoodard replied:
I updated this per our discussion. PTAL.

sir-buckyball commented on aip/general/0193.md:
```diff
@@ -62,11 +62,90 @@ a [`PreconditionFailure`][PreconditionFailure].
+
+1. The payload **must** contain an error message in the `message` field
+   found at the root of [`Status`][Status].
+1. The payload **should** contain
```
silly question: How does the API server know the appropriate locale to display to clients? While there may be existing patterns over HTTP (eg., the 'Accept-Language' header which is known to be wrong and inappropriately deferential to English), I'm not sure the same can be said for gRPC.

shwoodard replied:
iMO, it is beyond the scope of the AIP to prescribe how the infrastructure that a given company builds to implement the API designs found in these AIPs. Yes, 'Accept-Language' is reasonable.

sir-buckyball commented on aip/general/0193.md:
```diff
@@ -62,11 +62,90 @@ a [`PreconditionFailure`][PreconditionFailure].
+#### Localization
+
+The error message, found in the payload on `Status`, **must** be in
+English. If a localized error message is also required, the service
```
I realize this line isn't changed by you, but ick all the same. While a single company providing API's should adhere to a single language for consistency, as far as general API guidance goes I don't think it should necessarily be English.

shwoodard replied:
Agree.

toumorokoshi commented on aip/general/0193.md:
```diff
@@ -62,11 +62,90 @@ a [`PreconditionFailure`][PreconditionFailure].
-required, the service **should** use [`google.rpc.LocalizedMessage`][details]
-in the `details` field.
+The [`google.rpc.Status`][Status] error payload **should** contain two
+error messages.
```
nit: explain what those error message are here as a summary?

shwoodard replied:
This is outdated.

shwoodard committed:
pretty massive rewrite to address feedback

The biggest change is to relax the requirement around that the `message`
on `Status` **must not** change. Futhermore, the reasons and
requirements to use `LocalizedMessage` are clarified.

shwoodard committed:
remove bullets

shwoodard committed:
more edits

shwoodard committed:
even more edits

shwoodard committed:
bump the release date

shwoodard committed:
address pr feedback

shwoodard committed:
capitalize must

shwoodard committed:
change warning

shwoodard committed:
typo

shwoodard committed:
typos

noahdietz commented on aip/general/0193.md:
```diff
@@ -62,11 +62,109 @@ a [`PreconditionFailure`][PreconditionFailure].
+
+- is developer-facing, human-readable, and in the application's
+  native language
+- both explains the error and offer an actionable resolution to it
```
```suggestion
- both explains the error and offers an actionable resolution to it
```
Nit

noahdietz commented on aip/general/0193.md:
```diff
@@ -62,11 +62,109 @@ a [`PreconditionFailure`][PreconditionFailure].
+English rather than the application's native language so that messages
+are more easily universally searchable in common online knowledge bases.
+
+For new errors, the payload **must** include `ErrorInfo` and any
```
New errors as a new instance of an existing error type? Or for new error _types_?

maybe s/types/modes/ idk

shwoodard replied:
It's kind of left intentionally vague, but it is when you would introduce a new value for "reason" in `ErrorInfo`. Not, when you would add an `ErrorInfo` to an error with an existing error message and code pair, but an entirely new error scenario. Any help with the wording given that would be much appreciated.

shwoodard replied:
> When introducing a new error, a failure scenario that did not previously occur for the service, the payload **must** include `ErrorInfo` and any variables found in dynamic segments of the error message **must** be represented in `metadata`.

wdyt?

shwoodard replied:
I pushed this change. Feel free to "unresolve".

noahdietz commented on aip/general/0193.md:
```diff
@@ -62,11 +62,109 @@ a [`PreconditionFailure`][PreconditionFailure].
+
+Changes to `message` on `Status` **must not** change it if your
+service's API design governing body decides that to do so represents a
+breaking change and **should not** change it the text contains dynamic
```
I'm sorry I didn't grok this paragraph. Can this be reworded to be a bit more concise and separated? Something like,

```
If your service's API design governing body decides that changing a `Status.message` is a
breaking change, then the `Status.message` **must not** be changed.

If the `Status.message` contains dynamic data, but `ErrorInfo` was absent at the time of
launch, then the `Status.message` **should not** be changed.

To present a different, potentially more detailed error, consider using the
following approach.

// Go on to explain how that works...
```

Note: I used `Status.message`, but you can use your "message on Status" style too idc

shwoodard replied:
Done.

shwoodard committed:
more feedback addressed

shwoodard committed:
minor edits

shwoodard committed:
revision to #1098

shwoodard committed:
edits from live review with Noah

shwoodard committed:
fix sentence

shwoodard committed:
fix sentence

toumorokoshi commented on aip/general/0193.md:
```diff
@@ -62,11 +62,120 @@ a [`PreconditionFailure`][PreconditionFailure].
+- is a developer-facing, human-readable "debug message"
+- is presented in the service's native language
+- both explains the error and offers an actionable resolution to it
+  ([citation](https://cloud.google.com/apis/design/errors#error_model))
```
nit: it would be nice to consolidate from cloud.google.apis/design/ to here, vs link out. Maybe a future issue?

shwoodard replied:
Filed https://github.com/aip-dev/google.aip.dev/issues/1108

toumorokoshi commented on aip/general/0193.md:
```diff
@@ -62,11 +62,120 @@ a [`PreconditionFailure`][PreconditionFailure].
+language so that messages are easily searchable in common knowledge
+bases, such as StackOverflow&trade;.
+
+When introducing a new error, a failure scenario that did not previously
```
```suggestion
When introducing an error that represents a failure scenario that did not previously
```

nit: probably my  ignorance of english grammar, but "introducing" already implies new , and it seems like it's easier to read to explicitly state that it represents a failure scenario vs the comma.

toumorokoshi commented on aip/general/0193.md:
```diff
@@ -62,11 +62,120 @@ a [`PreconditionFailure`][PreconditionFailure].
+
+#### Changing error messages
+
+If your service's API design governing body decides that altering
```
these are Google AIP's, so we should know whether we govern this or not, correct? and state that clearly here.

shwoodard replied:
Some PAs have their own governance.

toumorokoshi replied:
I think this should be addressed - the guidance is vague and for someone arriving at this page, unclear as to how to figure out whether this is backwards incompatible or not.

Can it be explicitly stated that the change is allowed at minimum? the *must not* in here should be added by the PA in their own guidance then.

shwoodard replied:
I changed it so it explicitly states that `Status.message` **may** change if not disallowed. Is this what you're after?

toumorokoshi replied:
I'm more not a fan of the "governing bodies" which is vague and leaves a reader unable to figure out what to do next. 

If I was new to Google and I read this sentence, what would I do next? how would I find out if there was some other governing body?

I don't think we really need to discuss that this could or could not be a backwards incompatible change, unless we're going to say something authoritative like *should* or *should not*.

toumorokoshi commented on aip/general/0193.md:
```diff
@@ -62,11 +62,120 @@ a [`PreconditionFailure`][PreconditionFailure].
+
+<a name="status-message-warning"></a>
+**Warning:** **Use extreme caution** when changing `Status.message`.
+API consumers often treat this error message as **part of the API
```
nit: some of these feels like "Rationale" more than guidance. Guidance at most briefly elaborates on a particular choice, while rationale explains the choice in-depth.

shwoodard replied:
Ack, IMO it is important to include this warning here.

toumorokoshi commented on aip/general/0193.md:
```diff
@@ -62,11 +62,120 @@ a [`PreconditionFailure`][PreconditionFailure].
+message by adding [`google.rpc.LocalizedMessage`][LocalizedMessage] to
+[`Status.details`][Status details].
+
+The error message presented in `LocalizedMessage.message` **may** be the
```
is this sentence stating something meaningful? if it *may* do something and *may* not do it, then that's all the options available and this is basically saying either is acceptable. Which feels similar to not including specific guidance at all.

toumorokoshi replied:
if it is relevant: is it possible to clarify why one would choose one over the other?

shwoodard replied:
See below. It is explained.

toumorokoshi reviewed (commented):
sorry, half done but have a few comments. I'll add a few more in a little bit.

shwoodard committed:
Accept suggestion from @toumorokoshi

Co-authored-by: Yusuke Tsutsumi <tsutsumi.yusuke@gmail.com>

shwoodard committed:
fix line length

toumorokoshi commented on aip/general/0193.md:
```diff
@@ -62,11 +62,120 @@ a [`PreconditionFailure`][PreconditionFailure].
+  from the "debug message".
+- Ongoing, iterative error message improvements are expected.
+
+When including `LocalizedMessage`, both fields, `locale` and `message`,
```
can you include a proto example here? that'll help make it a little easier to follow the guidance.

shwoodard replied:
Do you mean a JSON example? Do you really think that is necessary? The proto has two string fields, `locale` and `message`. `locale` is like "en-US", stated and obvious and `message` has the error message. Also, the proto is linked to more than once.

toumorokoshi replied:
yeah, I guess it's a json payload.

I think an example at the top that has a fully populated response would be great, including:

- status
- code
- errorinfo (domain, reason metadata)
- localizedmessage (locale, message)

So one can see how to properly construct a response and how these values could related to each other.

It's hard to grasp how all of these values fit with each other a visualization to connect them all.

shwoodard replied:
Added. PTAL

toumorokoshi commented on aip/general/0193.md:
```diff
@@ -62,11 +62,120 @@ a [`PreconditionFailure`][PreconditionFailure].
+When adding an error message via `LocalizedMessage`, `ErrorInfo`
+**must** be introduced either before or at the same time. If there are
+dynamic segments found in the text, the values of these variables
+**must** be included `ErrorInfo.metadata`.  See, "[Dynamic
```
```suggestion
**must** be included in `ErrorInfo.metadata`.  See, "[Dynamic
```

toumorokoshi reviewed (changes requested):
sorry, a couple comments but very close.

One conversation around vague must / must not on backwards compatibility of status.message.

And I think we definitely need an example of a well-formed, fully populated error to put this guidance into context.

shwoodard committed:
Update aip/general/0193.md

Co-authored-by: Yusuke Tsutsumi <tsutsumi.yusuke@gmail.com>

shwoodard committed:
explicitly state that Status.message may change

shwoodard committed:
remove part about governing api design body

shwoodard committed:
add a sample json representation of an error response

shwoodard committed:
add jump link

shwoodard committed:
remove superfluous space

toumorokoshi reviewed (approved):
Thank you for all the hard work! LGTM.
